\documentclass{article}


\usepackage{enumerate}

\usepackage{graphicx}
\usepackage{algorithm}
\usepackage{algorithmic}
\usepackage{multirow}
\usepackage{amsmath}
\usepackage{color}
\newcommand{\SWITCH}[1]{\STATE \textbf{switch} (#1)}
\newcommand{\ENDPWITCH}{\STATE \textbf{end switch}}
\newcommand{\CASE}[1]{\STATE \textbf{case} #1\textbf{:} \begin{ALC@g}}
\newcommand{\ENDCASE}{\end{ALC@g}}
\newcommand{\CASELINE}[1]{\STATE \textbf{case} #1\textbf{:} }
\newcommand{\DEFAULT}{\STATE \textbf{default:} \begin{ALC@g}}
\newcommand{\ENDDEFAULT}{\end{ALC@g}}
\newcommand{\DEFAULTLINE}[1]{\STATE \textbf{default:} }
\newtheorem{mydef}{Definition}
\newtheorem{mylm}{Proposition}

\renewcommand{\algorithmicrequire}{\textbf{Input:}}
\renewcommand{\algorithmicensure}{\textbf{Output:}}


\begin{document}


\title{An Efficient Algorithm for Multiple Longest Common Subsequences
  (MLCS) Problem Based on Path Marking}

\author{Zhan Peng, Yuping Wang\footnote{Corresponding author}~,}


\maketitle

%begin{history}
%received{(Day Month Year)}
%revised{(Day Month Year)}
%accepted{(Day Month Year)}
%comby{(xxxxxxxxxx)}
%end{history}

\begin{abstract}

  Searching for the Multiple Longest Common Subsequences (MLCS) of
  multiple sequences is a classical NP-hard problem which is widely
  used in many areas such as bioinformatics and biomedicine,
  etc. Although significant efforts have been made to address this
  problem, the increasing amount of biological data requires more
  efficient methods. In this paper, we present a novel fast algorithm
  --- PM-MLCS for the MLCS problem, which is based on the dominant
  point approach and employs a new technique called path
  marking. During the construction of the dominant point graph, the
  path marking technique can real-timely mark the longest path from
  the source node to the current node, and once the graph has been
  constructed the longest paths corresponding to every MLCS can be
  found immediately. The experimental results show that, comparing
  with the expensive ``Minima'' operation that is widely used in many
  existing algorithms, the path marking is much more efficient, and
  therefore our algorithm performs better than those algorithms.

\end{abstract}

\textbf{multiple longest common subsequences (MLCS); Longest common
  subsequence (LCS); dominant point method}

% due to the effectiveness of the filter and the compactness of the AMT
% whether there are patterns starting at that position

\section{Introduction}
\label{sec:introduction}


Many kinds of biological data can be represented as a sequence of
symbols. For instance, proteins are sequences over 20 different
symbols (amino acids), and DNA (genes) can be represented as sequences
over four symbols A, C, G and T corresponding to the four bases.
Measuring the similarity of biological sequences is a fundamental
problem in bioinformatics, which can be used in many applications such
as cancer diagnoses, detection of the species’ common origin [2],
etc. One of the most important ways to measure the similarity of
biological sequences is to find their Multiple Longest Common
Subsequences (MLCS), which has been proved to be an NP-hard problem
\cite{Maier1978}. Traditionally, the works are focusing on either the
simplest case of two sequences, also known as the Longest Common
Subsequences (LCS) problem, or the problem’s special case of three
sequences. Although many algorithms have been proposed for the general
case of the problem, they are not very efficient in practice for
large-size data with many sequences due to high time and space
complexity. Particularly, with the successful implementation of the
human genome project and the development of the next-generation
sequencing techniques, the amount of biological sequences (as well as
other types of sequences from various fields) are growing explosively
[3]. Therefore, it is very important and valuable to design more
efficient MLCS algorithm to process large amount of biological (and
other types) sequences.

In this paper, we proposed a fast algorithm --- PM-MLCS for finding
MLCS of many sequences. PM-MLCS is based on the dominant point model,
in which the dominant points produced during processing are organized
to form a Directed Acyclic Graph (DAG), and each of the longest paths
from the source node to the end node corresponds to a MLCS. Different
from the other dominant point based algorithms that adopt a
time-consuming \emph{Minima} operation to eliminate the redundant
nodes, and then search for the longest paths from the source node to
the end node to find the MLCS, PM-MLCS employs a technical called
\emph{path marking} which has the ability that whenever a node is
generated, the longest path from the source node to the new node is
updated immediately, and once the DAG is constructed, all the MLCS can
be found directly from the end node back to the source node by the
marked path. The experimental results show that the \emph{path
  marking} technique is much efficient than the \emph{Minima}
operation, so that our algorithm is much faster than the compared
algorithms.

This paper is organized as follows: Section \ref{sec:related works}
introduces some preliminaries and reviews the related works. Section
\ref{sec:notations} lists the notations and terminology used in this
paper. In section \ref{sec:algorithm}, the our new algorithm is
presented in details. In section \ref{sec:experiments}, experimental
results comparing the performances of the new algorithms with existing
ones are presented. In Section \ref{sec:conclusion} we summarize the
paper.

\section{Preliminaries and Related Works}
\label{sec:related works}

In this section, we define some notations and terminologies used in
this paper and review some related works in the MLCS problem.

First of all, let $\Sigma$ denote a finite alphabet for the sequences.
For example, for the DNA sequences, $\Sigma=\{A, C, T, G\}$.

\textbf{Definition 1.} Let $s=c_1c_2...c_n$ be a sequence of length
$n$ over $\Sigma$, and the $i$-th character of $s$ is denoted by
$s[i]=c_i$. A \emph{subsequence} of $s$ is a sequence that can be
obtained by deleting zero or more (not necessarily consecutive)
characters from $s$. Formally, $s'=c_{i_1}c_{i_2}...c_{i_k}$ is called
a subsequence of $s$, if there is $1 \leq i_j \leq n$
($1 \leq j \leq k$), and for all $r$ and $t$ such that
$1 \leq r < t \leq k$,
there is $i_r < i_t$.\\

\textbf{Definition 2.} Given $d$ sequences $s_1, s_2, ..., s_d$ over
$\Sigma$, a Longest Common Subsequence (\emph{LCS}) of the $d$
sequences is a sequence $lcs$ such that: (1) $lcs$ is a subsequence of
all $s_i$ ($1 \leq i \leq d$) (2) there is no sequence satisfying (1)
and longer than $lcs$.\\

Generally, there exists more than one LCS for multiple sequences. For
example, given three DNA sequences $s_1 = ACTAGTGC$, $s_2 = TGCTAGCA$
and $s_3 = CATGCGAT$, there exists two LCSs of length 4, which are
$lcs_1 = CAGC$ and $lcs_2 = CTGC$, respectively. The MLCS problem is
to find all LCSs of three or more sequences. For the special case of
two sequences, the problem is usually called the LCS
problem. Particularly, specific algorithms developed for LCS problem
are generally inefficient for
MLCS problem.\\

Because of the importance of MLCS problem, several efficient
algorithms have been proposed in the past decades. Roughly, there are
two classes of the MLCS algorithms: the dynamic programming based
approaches and the dominant point based approaches.

\subsection{Dynamic Programming Based Approaches}
\label{sec:Dynamic Programming}

The classical approaches for the MLCS problem are based on dynamic
programming \cite{Smith1981}, \cite{Sankoff1972}. Given $d$ sequences:
$s_1,\, s_2,\,...,\, s_d$ of length $n_1,\, n_2,\, ...,\, n_d$
respectively, these approaches recursively construct a score table $T$
with $n_1 \times n_2 \times ... \times n_d$ cells, in which
$T[i_1,\, i_2,\, ...,\, i_d]$ records the length of the LCSs of the
prefixes $s_1[1...i_1]$, $s_2[1...i_2]$, ..., $s_d[1...i_d]$.
$T[i_1,\, i_2,\, ...,\, i_d]$ can be computed recursively by the
following formula:

\begin{equation*}
  T[i_1,\, i_2,\, ...,\, i_d] = 
  \begin{cases}
    0 & \text{if $\exists j(1 \leq j \leq d), i_j = 0$}\\
    T[i_1-1,\, ...,\, i_d-1] + 1  & \text{if $s_1[i_1] = s_2[i_2] =
      ... = s_d[i_d]$}\\
    max(\bar{T}) & \text{otherwise}
  \end{cases}
\end{equation*}

where
$\bar{T} = \{T[i_1-1,\, i_2,\, ...,\, i_d],\, T[i_1,\, i_2-1,\, ...,\,
i_d],\, ...,\, T[i_1,\, i_2,\, ...,\, i_d-1]\}$. Once the score table
$T$ is constructed, the MLCS can be collected by tracing back from the
last cell $T[n_1,\, n_2,\, ...,\, n_d]$ to the first cell
$T[0,\, 0,\, ...,\, 0]$. Fig. \ref{fig:DM} (a) shows an example of the
score table $T$ for two sequences $s_1 = ACTAGCTA$ and
$s_2 = TCAGGTAT$. The MLCS of these two sequences, which are $TAGTA$
and $CAGTA$, can be found by tracing back from $T[8,\, 8]$ to
$T[0,\, 0]$, as shown in Fig. \ref{fig:DM} (b).


\begin{figure}[htbp]
  \centering
  \includegraphics[width=0.9\textwidth]{./eps/DM}
  \caption{(a) The score table of the two DNA sequences ACTAGCTA and
    TCAGGTAT. (b) Constructing the MLCS, where the shaded cells
    conspond to dominant points.}
  \label{fig:DM}
\end{figure}

Obviously, the time and space complexity of dynamic programming based
approaches for a MLCS problem with $d$ sequences of length $n$ are
both $O(n^d)$ \cite{Hsu1984}, therefore, those approaches are not
practical for MLCS problems with more than two sequences. To reduce
the complexity of dynamic programming based approaches, various
methods have been proposed \cite{Hirschberg1977},
\cite{Apostolico1992}, \cite{Masek1980},
\cite{Rick1994}. Unfortunately, these methods primarily address the
LCS problem of two sequences.

\subsection{Dominant Point Based Approaches}
\label{sec:Dominant Point}

Many works have been proposed to improve dynamic programming based
approaches in order to process large size data, among which the
dominant point approaches are the most efficient. Before discussing
the dominant point based approaches, some definitions are introduced
first:

\textbf{Definition 3.} Given $d$ sequences $s_1,\, s_2,\, ...,\, s_d$
over $\Sigma$, a vector $p = (p_1,\, p_2,\, ...,\, p_d)$ is called a
\emph{match point} of the $d$ sequences, if
$s_1[p_1] = s_2[p_2] = ... = s_d[p_d] = \delta$, where $\delta$ is the
character corresponding to
the matched point $p$ and is denoted by $C(p)$.\\

\textbf{Definition 4.} Given two match points
$p = (p_1,\, p_2,\, ...,\, p_d)$ and $q = (q_1,\, q_2,\, ...,\, q_d)$
of $d$ sequences, we say that $p = q$, if and only if $p_i = q_i$
($1 \leq i \leq d$). On the other hand, if $p_i \leq q_i$
($1 \leq i \leq d$), we say $p$ \emph{dominates} $q$, which is denoted
by $p \preceq q$. Moreover, if $p_i < q_i$ ($1 \leq i \leq d$), we say
$p$ \emph{strongly dominates} $q$, which is denoted by $p \prec q$.
If $p \prec q$ and there is no match point $r$ such that
$p \prec r \prec q$ and $C(q) = C(r)$, we call $q$ a \emph{successor}
of $p$, or $p$ a \emph{precursor} of $q$. In particular, one match
point can have at most $|\Sigma|$ successors with each successor
corresponds to one character in
$\Sigma$. \\

\textbf{Definition 5.} The \emph{level} of a match point
$p = (p_1,\, p_2,\, ...,\, p_d)$ is defined to be
$L(p) = T[p_1,\, p_2,\, ...,\, p_d]$, where $T$ is the score table
constructed by formula []. A match point $p$ is called a
\emph{k-dominant point} (\emph{k-dominant} for short) if and only if:
(1) $L(p) = k$ (2) there is no other match point $q$ such that:
$L(q) = k$ and $q \preceq p$. All the $k$-dominants form a set denoted
by $D^k$.

The motivation of the dominant point based approaches is to reduce the
time and space complexity of the basic dynamic programming
methods. The key idea is based on the observation that only the
dominant points contribute to the construction of the MLCS (as shown
in Fig \ref{fig:DM} (b), the shaded cells correspond to the dominant
points).  Since the number of dominant points can be much smaller than
that of all cells in the score table $T$, a dominant point approach
that only identifies the dominant points, without filling the whole
score table, can greatly reduce the time and space complexity.

The search space of the dominant point based approaches can be
organized into a Directed Acyclic Graph (DAG): a node in DAG
represents a match point, while an edge $\langle p,\, q \rangle$ in
DAG represents that $q$ is a successor of $p$, i.e., $p \prec q$ and
$L(q) = L(p) + 1$.  Initially, the DAG contains only a \emph{source
  node} $(0,\, 0,\, ...,\, 0)$ with no incoming edges and an \emph{end
  node} $(\infty,\, \infty,\, ...,\, \infty)$ with no outgoing
edges. Then, the DAG can be constructed level by level as follows: at
first, let the level $k = 0$, and $D^0 = \{(0,\, 0,\, ...,\,
0)\}$. Next, with a forward iteration procedure, the $(k+1)$-dominants
$D^{k+1}$ are computed based on the $k$-dominants $D^k$, and this
procedure is denoted by $D^k \rightarrow D^{k+1}$. Specifically, each
node in $D^k$ is expanded by generating all its $|\Sigma|$ successors,
then a pruning operation called $Minima$ is performed to identify
those successors who dominant others, and only those dominants are
reserved to form $D^{k+1}$. Once all the nodes have been expanded, the
whole DAG is constructed, in which a longest path from the source node
to the end node corresponds to a LCS, thus, the MLCS problem becomes
finding all the longest paths from the source node to the end node.
Fig. \ref{fig:DAG} shows the DAG of the two DNA sequences $ACTAGCTA$
and $TCAGGTAT$ constructed by the general dominant based
algorithms. The black nodes in the DAG are duplicated points, while
the gray nodes in each level are non-dominant points, and both of
these nodes will be eliminated by the \emph{Minima} operation. The
main drawbacks of the dominant point based approaches are that a lot
of redundant nodes will be generated and the \emph{Minima} operation
needs to compare among the match points dimension by dimension, which
is very time consuming.


\begin{figure}[htbp]
  \centering
  \includegraphics[width=0.6\textwidth]{./eps/DAG}
  \caption{The DAG of the two sequences $ACTAGCTA$ and $TCAGGTAT$
    constructed by the general dominant point based algorithms, in
    which the black and gray nodes will be eliminated by the
    \emph{Minima} operation.}
  \label{fig:DAG}
\end{figure}

Hunt \cite{Hunt1977} proposed the first dominant point based algorithm
for two sequences with time complexity $O((r+n)log^n)$, where $r$ is
the number of nodes in DAG and $n$ is the length of the two
sequences. Afterwards, to further improve the efficiency, a variety of
dominant point based LCS/MLCS algorithms have been presented. Korkin
\cite{Korkin2001} proposed the first parallel MLCS algorithm with time
complexity $O(|\Sigma|d)$, where $d$ is the number of sequences. Chen
\cite{Chen2006} presented an efficient MLCS algorithm -- FAST-LCS for
DNA sequences, it introduced a novel data structure called successor
table to obtain the successors of nodes in constant time and used a
pruning operation to eliminate the non-dominant nodes in each
level. Wang \cite{Wang2011} improved the FAST-MLCS algorithm by using
a divide-and-conquer strategy to eliminate the non-dominant nodes,
which is very suitable for parallelization, it indicated that the
parallelized algorithm Quick-DPPAR had gained a near-linear speedup
compared to its serial version. Li \cite{Li2012} and Yang
\cite{Yang2010} made efforts to develop efficient parallel algorithms
on GPUs for the LCS problem and on cloud platform for the MLCS
problem, respectively. Unfortunately, Yang \cite{Yang2010} is not
suitable for the MLCS problem with multiple sequences due to the large
synchronous costs. Recently, Li \cite{LiICDE} and Li \cite{LiSIGKDD}
proposed two algorithms: PTop-MLCS and RLP-MLCS based on dominant
points, these algorithms used a novel graph model called Non-redundant
Common Subsequence Graph (NCSG) which can greatly reduce the redundant
nodes during processing, and adopted a two-passes topological sorting
procedure to find the MLCS. The authors claimed that the time
complexity of their algorithms is linear to the number of nodes in
NCSG.

In practice, for MLCS problems with large number of sequences, the
traditional algorithms usually need a long time and large space to
find the optimal solution (the real MLCS), to address this,
approximate algorithms have been investigated to quickly produce a
suboptimal solution (partial MLCS) and gradually improve it when given
more time, until an optimal one is found. Yang \cite{Yang2013}
proposed an approximate algorithm Pro-MLCS as well as its efficient
parallelization based on the dominant point model. Pro-MLCS can find
an approximate solution quickly, which only takes around 3 percent of
the entire running time, and then progressively generate better
solutions until obtaining the optimal one. Recently, Yang
\cite{Yang2014} proposed another two approximate algorithms SA-MLCS
and SLA-MLCS. SA-MLCS used an iterative beam widening search strategy
to reduce space usage during the iterative process of finding better
solutions. Based on SA-MLCS, SLA-MLCS, a space-bounded algorithm, is
developed to avoid space usage from exceeding available memory.


\section{PM-MLCS: A New Fast MLCS Algorithm}
\label{sec:PM-MLCS}

In this section, we present a new algorithm -- PM-MLCS to find the
MLCS of multiple sequences. The new algorithm is based on dominant
points model and a novel technique called path marking. Before
describing its detailed procedure, some data structures used are
introduced first.

\subsection{Data Structures}
\label{sec:data structures}

\subsubsection{Successor Table}
\label{sec:successor table}

One of the fundamental tasks during constructing the DAG is to
efficiently generate the successors of every node. To achieve this,
before constructing the DAG, a \emph{successor table} \cite{Chen2006}
is built for each sequence to enable the successors of each node be
generated in constant time. Given a sequence $s=c_1c_2...c_n$, its
corresponding successor table (denoted by $ST$) is a two-dimensional
array with $|\Sigma| \times (n+1)$ elements, in which the element of
$i$-th row and $j$-th column in $ST$ (denoted by $ST[i, j]$) is
defined as follows:
$$ST[i,j]=min\{m\;|\;c_m=\sigma_i,\; m \geq 1, m > j,\; 1 \leq i \leq
|\Sigma|,\; 0 \leq j \leq n\}$$ where $\sigma_i$ is the $i$-th
character in $\Sigma$. In fact, $ST[i,j]$ specifies the position of
the first occurrence of character $\sigma_i$ in $s$, starting from the
$(j+1)$-st position. For instance, the successor tables of the two DNA
sequences $ACTAGCTA$ and $TCAGGTAT$ are shown in Fig.\ref{fig:ST} (a)
and Fig.\ref{fig:ST} (b) respectively. Obviously, given $d$ sequences
of length $n$, all the corresponding successor tables can be built in
$O(d|\Sigma|n)$ time.


\begin{figure}[htbp]
  \centering
  \includegraphics[width=0.8\textwidth]{./eps/ST}
  \caption{(a) The successor table of the sequence ACTAGCTA. (b) The
    successor table of the sequence TCAGGTAT.}
  \label{fig:ST}
\end{figure}


By referring to the successor tables, the successors of one node can
be generated in constant time. For instance, the successors of the
node $(2, 2)$ in Fig.\ref{fig:DAG} are
$(T[1, 2],\;T[1, 2]) = (4, 3)$,\, $(T[2, 2],\;T[2, 2]) = (6, -)$,\,
$(T[3, 2],\;T[3, 2]) = (5, 4)$ and $(T[4, 2],\;T[4, 2]) = (4, 3)$,
corresponding to $A$, $C$, $G$, and $T$ respectively, where $(6, -)$
does not exist, which means $(2, 2)$ has no successor corresponding to
$C$, and actually a node can have no successors at all. Obviously,
given $d$ sequences, all the successors of one node can be generated
in $O(|\Sigma|d)$ time.

\subsubsection{The Structure of the Node in DAG}
\label{sec:Node}

For each node, say $t$, in DAG, it contains the following kinds of
information:

\begin{itemize}
\item The match point vector corresponding to $t$, which is also the
  unique identifier of $t$.
\item $D(t)$: the \emph{length} of the longest path(s) from the source
  node to $t$.
\item $Suc(t)$: all the \emph{successors} of $t$, i.e.,
  $Suc(t) = \{s\;|\; s$ is a successor of $t\}$.
\item $Pre(t)$: all the \emph{precursors} of $t$ which are in the
  longest paths from the source node to $t$, i.e.,
  $Pre(t) = \{p\;|\; t \in Suc(p)$ and $D(p) = D(t) - 1$\}. These
  precursors are called \emph{key precursors}.
\end{itemize}

Note that, for $Pre(t)$, not all precursors of $t$ are kept, instead,
only those key precursors which are in the longest paths from the
source node to $t$ are reserved. By tracing the key precursors from
the end node to source node, all the longest paths can be traversed
backward, accordingly, all the MLCS can be obtained. Further, for
$Suc(t)$ and $Pre(t)$, only the pointers to the corresponding nodes
are kept.

\subsubsection{Global Data Structures}
\label{sec:auxiliary}

\begin{itemize}
\item $DAG$ : the data structure to maintain all the generated nodes.
\item $Q$ : a first-in-first-out queue to store the nodes for
  expanding.
\end{itemize}

In our implementation, the $DAG$ is implemented as $|\Sigma|$ number
of hash tables with each table corresponds to one character in
$\Sigma$.  Whenever a new node (successor) $t$ is generated, it is
kept in the hash table corresponding to $C(t)$.

Another data structure $Q$ is used to keep the nodes for expanding.
Whenever a new node is created, it is inserted into the end of $Q$,
while the node at beginning of $Q$ is the one to be expanded next.

\subsection{The Procedure of the PM-MLCS}
\label{sec:PM-MLCS}

In general, the procedure of PM-MLCS can be divided into two phases:
in the first phase, the DAG is constructed and all the longest paths
from the source node to the end node are marked in the meantime; in
the second phase, all the longest paths are traversed backward from
the end node to the source node, and all the corresponding MLCS are
obtained.

Initially, the DAG contains only the source node and the end node.
Then from the source node, the DAG is constructed level by level by
repeatedly expanding the newly created nodes. Suppose the node $t$ is
being expanded, and one of the successors of $t$, say $s$, (actually
the match point vector of $s$) is obtained by referring to the
successor table. Depending on whether $s$ has been created before
(whether there exists a node already containing the match point
vector), different operations are carried out to construct the DAG:

\begin{enumerate}
\item If $s \notin DAG$, the following operations are performed:
  \begin{enumerate}
  \item $DAG \leftarrow DAG \cup \{s\}$
  \item $Suc(t) \leftarrow Suc(t) \cup \{s\}$
  \item $Pre(s) \leftarrow \{t\}$
  \item $D(s) \leftarrow D(t) + 1$
  \item Insert $s$ to the end of $Q$.
  \end{enumerate}

  That is, if $s$ is a new node which has not been created before, $s$
  will be created first and then incorporated into $DAG$. Next, the
  successor/key precursor relationships between $t$ and $s$ are
  constructed. Obviously, the length of the longest path from the
  source node to $s$ is equal to $D(t)+1$. Then, $s$ will be inserted
  to $Q$ for further expanding. Fig.\ref{fig:case1} shows this case,
  where the wavy arrow indicates the paths from the source node to
  $t$, and the straight/dashed arrows indicate the successor/key
  precursor relationships, respectively.

\begin{figure}[htbp]
  \centering
  \includegraphics[width=0.8\textwidth]{./eps/case1}
  \caption{If $s$ has not been created before, create it first and
    then insert it to DAG. }
  \label{fig:case1}
\end{figure}
  
\item If $s \in DAG$, i.e., $s$ has already been created before, it
  needs not to be created any more. Depending on the relation between
  $D(t)+1$ and $D(s)$, different update operations will be carried
  out:

\begin{figure}[htbp]
  \centering
  \includegraphics[width=1.0\textwidth]{./eps/3cases}
  \caption{The update operations for node $t$ and its successor $s$,
    which is based on the relation between $D(t)+1$ and $D(s)$.}
  \label{fig:3cases}
\end{figure}

%$Source ~ Node \rightarrow ... \rightarrow t \rightarrow s$
\begin{enumerate}[i]
  \item If $D(t)+1 < D(s)$, $Suc(t) \leftarrow Suc(t) \cup \{s\}$ is
    performed. This means that if the paths from the source node
    through $t$ to $s$ are shorter than the current longest paths to
    $s$, the only thing needs to do is adding $s$ to the successors of
    $t$ (after that, the next successor of $t$ can be processed).
    Fig. \ref{fig:3cases} (a) shows this case, where $s$ has two key
    precursors ($a$ and $b$) as well as three successors ($c$, $d$ and
    $e$), for clarity, the other successors and key precursors of $t$
    are not depicted in the figure.

  \item If $D(t)+1 = D(s)$, perform
    $Suc(t) \leftarrow Suc(t) \cup \{s\}$ and
    $Pre(s) \leftarrow Pre(s) \cup \{t\}$. This case is that if there
    exists paths from the source node through $t$ to $s$ having the
    same length with the current longest path to $s$, all these paths
    will be the longest paths from the source node to
    $t$. Fig.\ref{fig:3cases} (b) shows this case, where $t$ is added
    to the key precursors of $s$.

  \item If $D(t)+1 > D(S)$, perform
    $Suc(t) \leftarrow Suc(t) \cup \{s\}$, $Pre(s) \leftarrow \{t\}$,
    $D(s) \leftarrow D(t) + 1$ and $Update(Suc(s))$ in order. This is
    the most complicated case -- if there exists paths from the source
    node through $t$ to $s$ longer than the current longest paths to
    $s$, the longest paths to $s$ will be updated accordingly. Fig
    \ref{fig:3cases} (c) shows this case, where the old key precursors
    of $s$ are deleted and $t$ becomes the new key precursor of
    $s$. Note that, since the longest paths to $s$ have been changed,
    the longest paths to the successors of $s$ may also be changed
    accordingly. Therefore, a $Update$ operation is performed on each
    successor of $s$, as shown in Fig. \ref{fig:update}.  Suppose $s$
    has three successors $c$, $d$ and $e$, and there are
    $D(s) < D(c)$, $D(s) = D(d)$ and $D(s) < D(e)$, respectively (note
    that, $D(S)$ has already been updated to $D(t)+1$). According to
    the three cases we have just discussed, no actions need be done to
    successor $c$; for successor $d$, add $s$ to its key precursors;
    for successor $e$, delete its old key precursors and let $s$ be
    its new key precursor, analogously, since the longest paths to $e$
    have been changed, the successors of $e$ should also be updated
    recursively, furthermore this updating procedure will be repeated
    until encountering the end node.

\begin{figure}[htbp]
  \centering
  \includegraphics[width=0.9\textwidth]{./eps/update}
  \caption{Update the successors of $s$. (a) Before updating. (b)
    After updating, the successors of $e$ needs to be updated
    recursively.}
  \label{fig:update}
\end{figure}

\end{enumerate}
\end{enumerate}

The pseudocode of constructing the DAG by our PM-MLCS is presented in
Algorithm \ref{alg:DAG}. The $source$ and $end$ denote the source node
and end node respectively, and the operations for updating node $t$
and its successor $s$ are wrapped to a recursive sub-procedure called
$Update(t, s)$. It is worth pointing out that if a node has no
successors, the end node is defined to be its
successor. Fig.\ref{fig:path_marking} (a) shows an example of the
constructed DAG for the two sequences $s_1 = ACTAGCTA$ and $s_2 =
TCAGGTAT$ by using Algorithm \ref{alg:DAG}. The dashed arrows point to
the key precursors, and the length of longest paths to each node is
shown below that node. Clearly, the length of the MLCS is equal to
$D(end)-1$.

Once the DAG has been constructed, by tracing the key precursors from
the end node to source node, all the longest paths can be traversed
backward. This traversing procedure can be done by a simple
depth-first walk from the end node, and the result longest paths are
shown in Fig.\ref{fig:path_marking} (b), obviously, the MLCS are
$CAGTA$ and $TAGTA$.

\begin{algorithm}
  \caption{PM-MLCS: Construction of the DAG}\scriptsize
  \label{alg:DAG}
  \begin{algorithmic}[1]
    \REQUIRE ~~\\
    The successor tables of the sequences.\\
    \ENSURE ~~\\
    The DAG with all longest paths marked.
    \STATE
    \STATE $Pre(source) \leftarrow \emptyset$, $Suc(source)
    \leftarrow \emptyset$, $D(source) \leftarrow 0$
    \STATE $Pre(end) \leftarrow \emptyset$, $Suc(end)
    \leftarrow \emptyset$, $D(end) \leftarrow 0$
    \STATE $DAG \leftarrow \{source, end\}$
    \STATE $Q \leftarrow \{source\}$
    \STATE
    \WHILE{$Q \neq \emptyset$}
    \STATE $t \leftarrow $ The node at the beginning of $Q$
    \FOR {each successor $s$ of $t$}
    \IF{$s \notin DAG$}
    \STATE $DAG \leftarrow DAG \cup \{s\}$
    \STATE $Suc(t) \leftarrow Suc(t) \cup \{s\}$
    \STATE $Pre(s) \leftarrow \{t\}$
    \STATE $D(s) \leftarrow D(t) + 1$
    \STATE Insert $s$ to the end of $Q$.
    \ELSE
    \STATE $Update(t, s)$ 
    \ENDIF
    \ENDFOR
    \IF{$t$ has no successor}
    \STATE $Suc(t) \leftarrow \{end\}$
    \STATE $Update(t, end)$
    \ENDIF 
    \ENDWHILE
    \STATE
    \STATE \textbf{Update(t, s)}
    \STATE
    \IF{$D(t)+1 < D(s)$}
    \STATE $Suc(t) \leftarrow Suc(t) \cup \{s\}$
    \ELSIF{$D(t)+1 = D(S)$}
    \STATE $Suc(t) \leftarrow Suc(t) \cup \{s\}$
    \STATE $Pre(s) \leftarrow Pre(s) \cup \{t\}$
    \ELSE
    \STATE $Suc(t) \leftarrow Suc(t) \cup \{s\}$
    \STATE $Pre(s) \leftarrow \{t\}$
    \STATE $D(s) \leftarrow D(t) + 1$
    \FOR{each successor $s'$ of $s$}
    \STATE $Update(s, s')$
    \ENDFOR
    \ENDIF
    
  \end{algorithmic}
\end{algorithm}

\begin{figure}[htbp]
  \centering
  \includegraphics[width=0.8\textwidth]{./eps/path_marking}
  \caption{(a) The constructed DAG for sequences ACTAGCTA and TCAGGTAT
    by Algorithm 1. (b) The longest paths from the end node to source
    node from which the MLCS can be obtained.}
  \label{fig:path_marking}
\end{figure}


\section{Experimental Results}
\label{sec:experiments}

In this section, we evaluate the performance of our \emph{PM-MLCS}
algorithm and three other state-of-the-art algorithms
\emph{Top\_MLCS}\cite{LiICDE}, \emph{Quick-DP} \cite{Wang2011} and
\emph{Fast\_LCS} \cite{Chen2006} on MLCS problems of various data
sizes.

\subsection{Experimental Setups}
\label{sec:setup}

In the experiments, all the algorithms are run on a Inspur Corporation
K1 800 high-performance key host with Intel Xeon E7-8870, 4 chip, 4
cores/chip (2.80 GHz) and 1TB memory in the High Performance Computing
Center of Xidian University. The operating system is GNU/Linux
(2.6.32) x86\_64, and all the algorithms are implemented with C/C++
compiled by \emph{gcc} along with option '-O2', according to their
papers respectively. The benchmark datasets used in the experiments
are real biological sequences from the \emph{NCBI} \cite{} and
\emph{DIP} \cite{} corpus.

\subsection{Evaluation under various numbers of sequences}
\label{sec:number}

Firstly, we evaluate the testing algorithms under various numbers of
sequences. There are 17 groups of DNA ($|\Sigma|=4$) as well as
protein sequences ($|\Sigma|=20$) for testing, with the number of
sequences in each group varying from 3 to 100. The sequences in all
groups have the same length 100. All the algorithms are independly run
10 times on each group of data and the mean running times are computed
and shown in Table \ref{} and Fig.\ref{}.





\section{Acknowledgments}

This work is supported by National Natural Science Foundation of China
(No.61472297) and the Fundamental Research Funds for the Central
Universities (BDZ021430).


\begin{thebibliography}{99}

\bibitem{Maier1978} D. Maier, “The Complexity of Some Problems on
  Subsequences and Supersequences”, J. ACM, vol.25, pp.322-336, 1978.


\bibitem{Smith1981} T.F. Smith and M.S. Waterman, “Identification of
  Common Molecular Subsequences,” J. Molecular Biology, vol.147,
  pp.195-197, 1981.

\bibitem{Sankoff1972} D. Sankoff, “Matching Sequences under
  Deletion/Insertion Con- straints,” Proc. Nat’l Academy Sciences
  USA, vol.69, pp.4-6, Jan. 1972.

\bibitem{Hsu1984} W.J. Hsu and M.W. Du, “Computing a Longest Common
  Subsequence for a Set of Strings,” BIT Numerical Math.,
  24(1):45-59. 1984.

\bibitem{Hirschberg1977} D.S. Hirschberg, “Algorithms for the Longest Common Subse-
quence Problem,” J. ACM, vol. 24, pp. 664-675, 1977

\bibitem{Apostolico1992} A. Apostolico, S. Browne, and C. Guerra,
  “Fast Linear-Space Com- putations of Longest Common Subsequences,”
  Theoretical Computer Science, vol. 92, no. 1, pp. 3-17, 1992.

\bibitem{Masek1980} W.J. Masek and M.S. Paterson, “A Faster Algorithm
  Computing String Edit Distances,” J. Computer System Sciences,
  vol. 20, pp. 18-31, 1980.

\bibitem{Rick1994} C. Rick, “New Algorithms for the Longest Common
  Subsequence Problem,” Technical Report No. 85123-CS, Computer
  Science Dept., Univ. of Bonn, Oct. 1994.
  
\bibitem{Hunt1977} J. W. Hunt and T. G. Szymanski. A fast algorithm
  for computing longest common subsequences.  Communications of the
  ACM, 20(5):350-353, 1977.

\bibitem{Korkin2001} D. Korkin. A new dominant point-based parallel
  algorithm for multiple longest common subsequence problem. Technical
  report, TR01-148, Univ. of New Brunswick, 2001.

\bibitem{Chen2006} Y. Chen, A. Wan, and W. Liu. A fast parallel
  algorithm for finding the longest common sequence of multiple
  biosequences. BMC Bioinformatics, 7(Suppl 4):S4, 2006.

\bibitem{Wang2011} Q. Wang, D. Korkin, and Y. Shang. A fast multiple
longest common subsequence (MLCS) algorithm.  Knowledge and Data
Engineering, IEEE Transactions on, 23(3):321-334, 2011.

\bibitem{Li2012} Y. Li, Y. Wang, and L. Bao. Facc: a novel finite
  automaton based on cloud computing for the multiple longest common
  subsequences search. Mathematical Problems in Engineering, 2012,
  2012.

\bibitem{Yang2010} J. Yang, Y. Xu, and Y. Shang. An efficient parallel
  algorithm for longest common subsequence problem on GPUs. In
  Proceedings of the World Congress on Engineering, volume 1, pp.
  499-504, 2010.

\bibitem{Yang2013} J. Yang, Y. Xu, and Y. Shang. A new progressive
  algorithm for a multiple longest common subsequences problem and its
  efficient parallelization. Parallel and Distributed Systems, IEEE
  Transactions on, 24(5):862-870, 2013.

\bibitem{Yang2014} J. Yang, Y. Xu, and Y. Shang. A space-bounded
  anytime algorithm for the multiple longest common subsequence
  problem. Knowledge and Data Engineering, Transactions on,
  26(11):2599-2609, 2014.

\bibitem{LiICDE} Yanni Li, Yuping Wang, Zhensong Zhang, etc. A Novel
  Fast and Memory Efficient Parallel MLCS Algorithm for Long and
  Large-Scale Sequences Alignments.  IEEE International Conference on
  Data Engineering (ICDE). pp.1170-1181. 2016.

\bibitem{LiSIGKDD} Li, Yanni and Li, Hui and Duan, Tihua, etc. A Real
  Linear and Parallel Multiple Longest Common Subsequences (MLCS)
  Algorithm. Proceedings of the 22nd ACM SIGKDD International
  Conference on Knowledge Discovery and Data
  Mining. pp.1725-1734. 2016.

\bibitem{NCBI} Available:
  http://www.ncbi.nlm.nih.gov/nuccore/110645304?report=fasta

\bibitem{DIP} Database of Interacting Proteins. Available:
  http://dip.doe-mbi.ucla.edu/dip/Download.cgi




\bibitem{pizzachil} {\it http://pizzachil.dcc.uchile.cl/}

\end{thebibliography}  


\vspace*{-0.01in}
%\vspace*{-0.3in}
\noindent
\rule{12.6cm}{.1mm}
pp
\end{document}

%%% Local Variables:
%%% mode: latex
%%% TeX-master: t
%%% End:

%%% Local Variables:
%%% mode: latex
%%% TeX-master: t
%%% End:
